### PR TITLE
Fix macaroon auth

### LINKED ISSUE #203: Can't connect to JAAS on master
#187 apparently wasn't tested against JAAS because there are a number of issues.  The proximate error is: https://paste.ubuntu.com/26546737/

In investigating this, however, there are some other issues currently being masked by this.

### PR DESCRIPTION
**Pull request by johnsca**
~~Depends on https://github.com/go-macaroon-bakery/py-macaroon-bakery/pull/46~~

Doesn't actually require that, though that fixes an edge case that users might hit.

This also has a drive-by possible fix for #172.

### CODE REVIEW

**Comment by johnsca** [comment]
So much for hopefully fixing `test_scp`.  :/

File: examples/connect_current_model.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+"""
+This example:
+
+1. Creates a model on the current controller
+2. Deploys a charm to it.
+3. Attempts to ssh into the charm
```
Comment (Cynerva) [inline examples/connect_current_model.py]: This docstring isn't accurate

File: juju/client/connection.py [hunk 2]
```diff
@@ -454,32 +463,44 @@ def connect_params(self):
...
+                response = result['response']
+                macaroonJSON = response.get('discharge-required')
                 if macaroonJSON is None:
-                    self.info = result['response']
+                    self.info = response
                     success = True
                     return result
                 # TODO implement macaroon authentication
-                raise NotImplementedError('macaroon authentication not implemented yet')
+                error_msg = response['discharge-required-error']
```
Comment (Cynerva) [inline juju/client/connection.py]: Can we safely assume a response that contains `discharge-required` will always contain `discharge-required-error` as well? I.E. we won't hit a `KeyError` here?

File: juju/client/connector.py [hunk 3]
```diff
@@ -104,12 +104,19 @@ def connection(self):
...
         # instead of just the first one.
         endpoint = controller['api-endpoints'][0]
-        models = self.jujudata.models()[controller_name]
-        account = self.jujudata.accounts()[controller_name]
+        accounts = self.jujudata.accounts()
+        if controller_name not in accounts:
+            raise JujuAuthError('Account not found for controller: {}'.format(
+                controller_name))
+        account = accounts[controller_name]
+        models = self.jujudata.models().get(controller_name, {})['models']
```
Comment (Cynerva) [inline juju/client/connector.py]: This looks a bit odd - if `self.jujudata.models()` doesn't container `controller_name`, then it'll default to `{}`, try to get `'models'` from it, and raise a `KeyError` yeah?

File: juju/client/connector.py [hunk 4]
```diff
@@ -104,12 +104,19 @@ def connection(self):
...
+        if model_name not in models:
+            raise JujuConnectionError('Model found: {}'.format(model_name))
```
Comment (Cynerva) [inline juju/client/connector.py]: Model not found?

**Comment by Cynerva** [comment]
Commented with a couple nits but nothing that really needs to be addressed. LGTM!

I can't merge PRs on this repo so someone else will have to do it.

File: juju/client/connection.py (hunk 2 above)
Comment (johnsca) [reply juju/client/connection.py]: That's correct, from what @rogpeppe told me.

**Comment by johnsca** [comment]
Closing in favor of #206

**Pull request closed without merging**